src/llama-kv-cells.h

```diff
@@ -0,0 +1,259 @@
+        }
+
+        return false;
+    }
+
+    // return true if the cell becomes empty (i.e. it did not contain seq_id before the call)
+    bool seq_keep(uint32_t i, llama_seq_id seq_id) {
+        assert(i < pos.size());
+
+        if (seq[i].test(seq_id)) {
+            seq[i].reset();
+            seq[i].set(seq_id);
+
+            return false;
+        }
+
+        if (seq[i].any()) {
+            seq[i].reset();
+            pos[i] = -1;
+
+            used--;
+
+            return true;
+        }
+
+        assert(pos[i] == -1);
+
+        return false;
+    }
+
+    bool seq_has(uint32_t i, llama_seq_id seq_id) const {
+        assert(i < pos.size());
+        assert(seq_id >= 0);
+
+        return seq[i].test(seq_id);
+    }
+
+    // note: call only if the cell is not empty
+    bool seq_add(uint32_t i, llama_seq_id seq_id) {
+        assert(i < pos.size());
+        assert(pos[i] != -1);
+
+        if (seq[i].none()) {
+            seq[i].set(seq_id);
+
+            used++;
+
+            return true;
+        }
+
+        return false;
+    }
+
+    // note: call only if the cell is not empty
+    llama_pos get_pos(uint32_t i) const {
+        assert(i < pos.size());
+        assert(pos[i] != -1);
+
+        return pos[i];
+    }
+
+    // note: call only if the cell is not empty
+    llama_pos get_delta(uint32_t i) const {
+        assert(i < pos.size());
+        assert(pos[i] != -1);
+
+        return delta[i];
+    }
+
+    bool pos_in(uint32_t i, llama_pos p0, llama_pos p1) const {
+        assert(i < pos.size());
+
+        return pos[i] >= p0 && pos[i] < p1;
+    }
+
+    // note: call only if the cell is empty
+    void pos_set(uint32_t i, llama_pos p) {
+        assert(i < pos.size());
+        assert(pos[i] == -1);
+
+        pos[i] = p;
+        used++;
+    }
+
+    // pos[i] = pos[i] + d
+    // note: call only if the cell is not empty
+    bool pos_add(uint32_t i, llama_pos d) {
+        assert(i < pos.size());
+        assert(pos[i] != -1);
+
+        pos[i]   += d;
+        delta[i] += d;
+
+        has_delta = true;
+
+        if (pos[i] < 0) {
+            pos[i] = -1;
+            seq[i].reset();
+
+            used--;
+
+            return true;
+        }
+
+        return false;
+    }
+
+    // pos[i] = pos[i] / d
+    // note: call only if the cell is not empty
+    void pos_div(uint32_t i, int d) {
+        assert(i < pos.size());
+        assert(pos[i] != -1);
+
+        const llama_pos p_old = pos[i];
+
+        pos[i]   /= d;
+        delta[i] += p_old - pos[i];
+
+        has_delta = true;
+    }
+
+    bool pos_has_shift() const {
+        return has_delta;
+    }
+
+    void pos_reset_delta() {
+        has_delta = false;
+
+        for (uint32_t i = 0; i < delta.size(); ++i) {
+            delta[i] = 0;
+        }
+    }
+
+private:
+    uint32_t used = 0; // used cells (i.e. at least one seq_id)
+
+    bool has_delta = false;
+
+    std::vector<llama_pos> pos;
+    std::vector<llama_pos> delta;
```

Comment:
I suggest renaming `has_delta` and `delta` to `has_shift` and `shift` for consistency, as well as adding a comment explaining the purpose. Also applies to `pos_reset_delta` and `get_delta`.

Reply:
Added a comment explaining the purpose in [44856a7](https://github.com/ggml-org/llama.cpp/pull/13706/commits/44856a70cd0e70213c424fa0072a150cd68b99b1).